csharp/ql/lib/experimental/code/csharp/MissingAntiForgeryTokenValidationAspNetCore.qll

```diff
@@ -0,0 +1,159 @@
+/** Library to support cs/web/missing-token-validation-aspnetcore query */
+
+private import csharp
+private import semmle.code.csharp.frameworks.system.Web
+private import semmle.code.csharp.frameworks.system.web.Helpers as Helpers
+private import semmle.code.csharp.frameworks.system.web.Mvc as Mvc
+private import semmle.code.csharp.frameworks.microsoft.AspNetCore as AspNetCore
+
+/** Defines classes used for anti-forgery tokens. */
+class AntiForgeryClass extends Class {
+  AntiForgeryClass() {
+    this instanceof Helpers::AntiForgeryClass or
+    this instanceof AspNetCore::AntiForgeryClass
+  }
+
+  /** Gets the `Validate` method. */
+  Method getValidateMethod() {
+    result = this.(Helpers::AntiForgeryClass).getValidateMethod() or
+    result = this.(AspNetCore::AntiForgeryClass).getValidateMethod()
+  }
+}
+
+/** Defines the anti-forgery token attribute. */
+class ValidateAntiForgeryTokenAttribute extends Attribute {
+  ValidateAntiForgeryTokenAttribute() {
+    this instanceof Mvc::ValidateAntiForgeryTokenAttribute or
+    this instanceof AspNetCore::ValidateAntiForgeryAttribute
+  }
+}
+
+/** Defines a generic controller including Mvc and AspNetCore */
+class Controller extends Class {
+  Controller() {
+    this instanceof Mvc::Controller or
+    this instanceof AspNetCore::MicrosoftAspNetCoreMvcController
+  }
+
+  /** Gets a Method with a POST action */
+  Method getAPostActionMethod() {
+    result = this.(Mvc::Controller).getAPostActionMethod()
+    or
+    exists(Method method |
+      method = this.(AspNetCore::MicrosoftAspNetCoreMvcController).getAnActionMethod() and
+      method.getAnAttribute() instanceof AspNetCore::MicrosoftAspNetCoreMvcHttpPostAttribute and
+      result = method
+    )
+  }
+}
+
+/** An `AuthorizationFilter` that calls the `AntiForgery.Validate` method. */
+class AntiForgeryAuthorizationFilter extends Mvc::AuthorizationFilter {
+  AntiForgeryAuthorizationFilter() {
+    this.getOnAuthorizationMethod().calls*(any(AntiForgeryClass a).getValidateMethod())
+  }
+}
+
+/** Find a filter that auto-validates anti-forgery tokens. */
+class AutoValidateAntiForgeryTokenFilter extends Expr {
+  AutoValidateAntiForgeryTokenFilter() {
+    exists(
+      MethodCall addControllers, LambdaExpr lambda, ParameterAccess options, PropertyCall filters,
+      MethodCall add
+    |
+      // "AddMvc", "AddControllersWithViews", "AddMvcCore", "AddControllers", "AddRazorPages", so generalised to "Add*" to future-proof
```

Comment:
If all the methods are in `Microsoft.Extensions.DependencyInjection` then maybe we should check the qualifier?